### ISSUE TITLE
Plugins on swarm with private repos

### ISSUE THREAD

**Issue by cpuguy83** [moby/moby#34010]
Now that plugins on swarm are in (#33575), we need a follow-up to add support for private repos.

In the process of getting swarm plugins in, it was decided to not carry-on the existing hack for passing through registry auth from the client. In doing so, it means that plugins on swarm can only work with public images.

After some discussion it was decided that using secrets to pass registry auth is the best way forward.

Currently swarm only distributes secrets by inspecting the `ContainerSpec` to determine what secrets are needed for a given task.
Since the plugin spec is opaque to swarm, swarm cannot just inspect the plugin spec to look for secrets.
So we need to be able to have swarm distributes secrets for a task without requiring it to look into a given runtime's spec.
Ideally I think the we can just list references to secrets in the `TaskTemplate` directly... swarm will still need to look at the `ContainerSpec` for backwards compat, but for plugins it can just look there.

Note: This also solves the issue of giving a plugin access to a secret.

With the secret distributed, believe we can handle the registry auth bit in the plugin spec (by specifying which secret is for registry auth).

**Comment by aaronlehmann** [comment]
> Ideally I think the we can just list references to secrets in the TaskTemplate directly... swarm will still need to look at the ContainerSpec for backwards compat, but for plugins it can just look at the task template.

Yes, I think this is the right approach. I would like to separate references to secrets and other resources (in the `TaskTemplate`, where the swarm manager can easily access them) from parameters that control the presentation of those secrets or other items (for example, where a secret is mounted inside a container or plugin).

There was some preliminary work on this here: https://github.com/docker/swarmkit/pull/2024/commits/4fbfb09cc0f9ba8be6f0ef3b02f24858e327c893
